Sources/SentryCrash/Recording/Tools/SentryCrashCxaThrowSwapper.c

```diff
@@ -0,0 +1,383 @@
+
+#include "SentryCrashCxaThrowSwapper.h"
+
+#include <dlfcn.h>
+#include <errno.h>
+#include <execinfo.h>
+#include <mach-o/dyld.h>
+#include <mach-o/nlist.h>
+#include <mach/mach.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/types.h>
+
+#include "SentryAsyncSafeLog.h"
+#include "SentryCrashMach-O.h"
+#include "SentryCrashPlatformSpecificDefines.h"
+
+typedef struct {
+    uintptr_t image;
+    uintptr_t function;
+} SentryCrashAddressPair;
+
+static cxa_throw_type g_cxa_throw_handler = NULL;
+static const char *const g_cxa_throw_name = "__cxa_throw";
+
+static SentryCrashAddressPair *g_cxa_originals = NULL;
+static size_t g_cxa_originals_capacity = 0;
+static size_t g_cxa_originals_count = 0;
+
+static void
+addPair(SentryCrashAddressPair pair)
+{
+    SENTRY_ASYNC_SAFE_LOG_DEBUG(
+        "Adding address pair: image=%p, function=%p", (void *)pair.image, (void *)pair.function);
+
+    if (g_cxa_originals_count == g_cxa_originals_capacity) {
+        g_cxa_originals_capacity *= 2;
+        g_cxa_originals = (SentryCrashAddressPair *)realloc(
+            g_cxa_originals, sizeof(SentryCrashAddressPair) * g_cxa_originals_capacity);
+        if (g_cxa_originals == NULL) {
+            SENTRY_ASYNC_SAFE_LOG_ERROR(
+                "Failed to realloc memory for g_cxa_originals: %s", strerror(errno));
+            return;
+        }
+    }
+    memcpy(&g_cxa_originals[g_cxa_originals_count++], &pair, sizeof(SentryCrashAddressPair));
+}
+
+static void
+sentrycrashct_clear_pairs(void)
+{
+    SENTRY_ASYNC_SAFE_LOG_TRACE("Clearing all stored pairs");
+
+    if (g_cxa_originals == NULL) {
+        SENTRY_ASYNC_SAFE_LOG_WARN("g_cxa_originals is NULL, nothing to clear");
+        return;
+    }
+
+    // Free the allocated memory
+    free(g_cxa_originals);
+    g_cxa_originals = NULL;
+    g_cxa_originals_count = 0;
+    g_cxa_originals_capacity = 0;
+
+    SENTRY_ASYNC_SAFE_LOG_TRACE("Freed memory for stored pairs");
+}
+
+static uintptr_t
+findAddress(void *address)
+{
+    SENTRY_ASYNC_SAFE_LOG_TRACE("Finding address for %p", address);
+
+    for (size_t i = 0; i < g_cxa_originals_count; i++) {
+        if (g_cxa_originals[i].image == (uintptr_t)address) {
+            return g_cxa_originals[i].function;
+        }
+    }
+    SENTRY_ASYNC_SAFE_LOG_WARN("Address %p not found", address);
+    return (uintptr_t)NULL;
+}
+
+static void
+__cxa_throw_decorator(void *thrown_exception, void *tinfo, void (*dest)(void *))
+{
+    const int k_requiredFrames = 2;
+
+    SENTRY_ASYNC_SAFE_LOG_TRACE("Decorating __cxa_throw");
+
+    if (g_cxa_throw_handler != NULL) {
+        g_cxa_throw_handler(thrown_exception, tinfo, dest);
+    }
+
+    void *backtraceArr[k_requiredFrames];
+    int count = backtrace(backtraceArr, k_requiredFrames);
+
+    Dl_info info;
+    if (count >= k_requiredFrames) {
+        if (dladdr(backtraceArr[k_requiredFrames - 1], &info) != 0) {
```

Comment:
@supervacuus I'm a bit unsure about this approach here. If we don't call the original `cxa_throw_type` apps crash when throwing C++ exceptions. It's very easy to reproduce. Basically swap cxa_throw, comment out `original(thrown_exception, tinfo, dest);` and then throw any C++ exception in a try catch. Then the app cashes. Do you have any idea what could be the motivation for only calling this if the `count >= k_requiredFrames`?